Adding B3 propagator

TomRoSystems opened the pull request.

TomRoSystems committed:
Adding B3 propagator

bogdandrutu commented on api/include/opentelemetry/trace/propagation/b3_propagator.h:
```diff
@@ -0,0 +1,278 @@
+#pragma once
+
+#include <array>
+#include <iostream>
+#include <map>
+#include <string>
+#include "opentelemetry/common/key_value_iterable.h"
+#include "opentelemetry/context/context.h"
+#include "opentelemetry/nostd/shared_ptr.h"
+#include "opentelemetry/nostd/span.h"
+#include "opentelemetry/nostd/string_view.h"
+#include "opentelemetry/nostd/variant.h"
+#include "opentelemetry/trace/default_span.h"
+#include "opentelemetry/trace/propagation/http_text_format.h"
+#include "opentelemetry/trace/span.h"
+#include "opentelemetry/trace/span_context.h"
```
Order of includes, use clang to format

TomRoSystems replied:
This PR is running same set of checks as opentelemtry-cpp and clang formatting is OK:
https://github.com/TomRoSystems/opentelemetry-cpp/pull/1/checks?check_run_id=1724608707

bogdandrutu commented on api/include/opentelemetry/trace/propagation/b3_propagator.h:
```diff
@@ -0,0 +1,278 @@
+namespace propagation
+{
+
+static const nostd::string_view B3CombinedHeader = "b3";
+
+static const nostd::string_view B3TraceIdHeader = "X-B3-TraceId";
+static const nostd::string_view B3SpanIdHeader  = "X-B3-SpanId";
+static const nostd::string_view B3SampledHeader = "X-B3-Sampled";
```
Do you need this to be public and exported?

bogdandrutu commented on api/include/opentelemetry/trace/propagation/b3_propagator.h:
```diff
@@ -0,0 +1,278 @@
+                             X-B3-TraceId                X-B3-SpanId    X-B3-ParentSpanId (ignored)
+*/
+
+static const int kTraceIdHexStrLength   = 32;
+static const int kSpanIdHexStrLength    = 16;
+static const int kTraceFlagHexStrLength = 1;
```
Same question for these values, if not we can move them to the cc file.

TomRoSystems replied:
This file was highly inspired by https://github.com/open-telemetry/opentelemetry-cpp/blob/880e392e10b6d8e964ade2e22d7231950d6d2443/api/include/opentelemetry/trace/propagation/http_trace_context.h and all API in C++ is header-only so there's no cc file.

bogdandrutu commented on api/include/opentelemetry/trace/propagation/b3_propagator.h:
```diff
@@ -0,0 +1,278 @@
+    return SpanContext::GetInvalid();
+  }
+
+  static TraceId GenerateTraceIdFromString(nostd::string_view trace_id)
+  {
+    uint8_t buf[kTraceIdHexStrLength / 2];
+    uint8_t *b_ptr = buf;
+    GenerateHexFromString(trace_id, kTraceIdHexStrLength, b_ptr);
+    return TraceId(buf);
+  }
+
+  static SpanId GenerateSpanIdFromString(nostd::string_view span_id)
+  {
+    uint8_t buf[kSpanIdHexStrLength / 2];
+    uint8_t *b_ptr = buf;
+    GenerateHexFromString(span_id, kSpanIdHexStrLength, b_ptr);
+    return SpanId(buf);
+  }
```
No need to be public correct?

TomRoSystems replied:
Same as with https://github.com/open-telemetry/opentelemetry-cpp/blob/880e392e10b6d8e964ade2e22d7231950d6d2443/api/include/opentelemetry/trace/propagation/http_trace_context.h - I think it's public for testing purposes.

bogdandrutu commented on api/include/opentelemetry/trace/propagation/b3_propagator.h:
```diff
@@ -0,0 +1,278 @@
+    return SpanId(buf);
+  }
+
+  static TraceFlags GenerateTraceFlagsFromString(nostd::string_view trace_flags)
+  {
+    if (trace_flags.length() != 1 || (trace_flags[0] != '1' && trace_flags[0] != 'd'))
+    {
+      return TraceFlags(0);  // check for invalid length of flags
+    }
+    return TraceFlags(TraceFlags::kIsSampled);  // treat 'd' as kIsSampled
+  }
```
Same.

TomRoSystems committed:
OTLP change default port number (#522)

seemk commented on api/include/opentelemetry/trace/propagation/b3_propagator.h:
```diff
@@ -0,0 +1,278 @@
+      }
+      else
+      {
+        buf[i / 2] += tmp;
+      }
+    }
+  }
+
+  // Converts a single character to a corresponding integer (e.g. '1' to 1), return -1
+  // if the character is not a valid number in hex.
+  static uint8_t HexToInt(char c)
+  {
+    if (c >= '0' && c <= '9')
+    {
+      return (int)(c - '0');
+    }
+    else if (c >= 'a' && c <= 'f')
+    {
+      return (int)(c - 'a' + 10);
+    }
+    else if (c >= 'A' && c <= 'F')
+    {
+      return (int)(c - 'A' + 10);
+    }
+    else
+    {
+      return -1;
+    }
+  }
+
+  static bool IsValidHex(nostd::string_view string_view)
+  {
+    for (int i = 0; i < string_view.length(); i++)
+    {
+      if (!(string_view[i] >= '0' && string_view[i] <= '9') &&
+          !(string_view[i] >= 'a' && string_view[i] <= 'f'))
+        return false;
+    }
+    return true;
+  }
```
Would it make sense to move these elsewhere? It's repeated here and at http_trace_context

seemk commented on api/include/opentelemetry/trace/propagation/b3_propagator.h:
```diff
@@ -0,0 +1,278 @@
+    auto singleB3Header = getter(carrier, B3CombinedHeader);
+    if (singleB3Header != "")
+    {
+      trace_id = singleB3Header.substr(0, kTraceIdHexStrLength);
```
B3 trace ID can be either 64 or 128 bits (that is 16 or 32 hex chars): https://github.com/openzipkin/b3-propagation#traceid-1

seemk commented on api/include/opentelemetry/trace/propagation/b3_propagator.h:
```diff
@@ -0,0 +1,278 @@
+  {
+    uint8_t buf[kTraceIdHexStrLength / 2];
+    uint8_t *b_ptr = buf;
+    GenerateHexFromString(trace_id, kTraceIdHexStrLength, b_ptr);
```
This would be a buffer overrun in case of 64 bit trace ids (which are unhandled currently, but a word of caution)

seemk commented on api/include/opentelemetry/trace/propagation/b3_propagator.h:
```diff
@@ -0,0 +1,278 @@
+      trace_id = singleB3Header.substr(0, kTraceIdHexStrLength);
+      if (singleB3Header.size() > 33)
+      {
+        span_id = singleB3Header.substr(33, kSpanIdHexStrLength);
```
What would happen if the input header size is 40?

seemk commented on api/include/opentelemetry/trace/propagation/b3_propagator.h:
```diff
@@ -0,0 +1,278 @@
+    char trace_flags[2];
+    TraceFlags(span_context.trace_flags()).ToLowerBase16(trace_flags);
+    // Note: This is only temporary replacement for appendable string
+    std::string hex_string = "";
```
This is on the hot path and will most likely allocate twice here when appending. Could use a stack buffer here with the appropriate size, since a string_view is passed to the setter anyway.

seemk commented on api/test/trace/propagation/b3_propagation_test.cc:
```diff
@@ -0,0 +1,138 @@
+  EXPECT_EQ(headers["X-B3-TraceId"], "0102030405060708090a0b0c0d0e0f10");
+  EXPECT_EQ(headers["X-B3-SpanId"], "0102030405060708");
+  EXPECT_EQ(headers["X-B3-Sampled"], "1");
+}
```
Would be nice to have a test of left-padding an incoming 64 bit trace id with zeroes

reyang committed:
Add spell check to CI (#527)

lalitb committed:
Update PR merge requirement. (#524)

lalitb committed:
fix namespace in trace_state class (#526)

seemk committed:
Fix grpc_cpp_plugin executable import (#525)

pyohannes committed:
Add a test service for the W3C validation harness (#487)

ThomsonTan committed:
Fix missing include dir for configure_package_config_file (#514)

pyohannes committed:
Rename SpanContext.HasRemoteParent to SpanContext.IsRemote (#529)

ThomsonTan committed:
Fix calling target_include_directories on library folder (#520)

lalitb committed:
Resource sdk Implementation (#502)

TomRoSystems committed:
B3 Propoagation:
- added support for TraceId 16 and 32 length
- added tests for scenario above and inject multi-header
nostd::string_view - added find for single character

TomRoSystems committed:
Fix formatting

TomRoSystems committed:
Make traceId left-padded

TomRoSystems committed:
Merge branch 'master' into b3_propagator

TomRoSystems committed:
Rename SpanContext.HasRemoteParent to SpanContext.IsRemote

TomRoSystems committed:
Fix for condition always false + test for invalid hex value

TomRoSystems committed:
Proper fix with latest master branch

reyang committed:
Move from master to main (step1) (#531)

reyang committed:
finish master to main migration (#533)

mtwo committed:
Added Zoom passcode (#532)

lalitb committed:
Sending sync request through http_client (#448)

ThomsonTan committed:
Fix install location of CMake config script (#528)

ThomsonTan committed:
Improve invalid span Id check (#538)

TomRoSystems committed:
Remove unused variables. Use constants for array sizes.

lalitb committed:
Merge branch 'main' into b3_propagator

TomRoSystems commented:
Already merged https://github.com/open-telemetry/opentelemetry-cpp/pull/523
